src/goto-instrument/replace_function_bodies.cpp

```diff
@@ -0,0 +1,367 @@
+/*******************************************************************\
+
+Module: Replace bodies of goto functions
+
+Author: DiffBlue Limited.
+
+\*******************************************************************/
+
+#include "replace_function_bodies.h"
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <memory>
+#include <sstream>
+#include <util/make_unique.h>
+#include <util/message.h>
+#include <langapi/language_util.h>
+
+class replace_function_bodiest
+{
+protected:
+  virtual void generate_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) = 0;
+
+public:
+  virtual ~replace_function_bodiest() = default;
+
+  void replace_function_body(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name)
+  {
+    function.body.clear();
+    generate_parameter_names(function, symbol_table, function_name);
+    generate_function_body(function, symbol_table, function_name);
+  }
+
+private:
+  /// Generate parameter names for unnamed parameters.
+  /// CBMC expect functions to have parameter names
+  /// if they are called and have a body
+  void generate_parameter_names(
+    goto_functiont &function,
+    symbol_tablet &symbol_table,
+    irep_idt function_name) const
+  {
+    namespacet ns(symbol_table);
+    int param_counter = 0;
+    for(auto &parameter : function.type.parameters())
+    {
+      if(
+        parameter.type().id() == ID_pointer &&
+        !is_constant_or_has_constant_components(parameter.type(), ns))
+      {
+        if(parameter.get_identifier().empty())
+        {
+          const std::string param_base_name =
+            parameter.get_base_name().empty()
+              ? "__param$" + std::to_string(param_counter++)
```

Comment:
To avoid collision you might consider using the cprover prefix (I suppose this doesn't matter as `$` is illegal in C variable names, and I think is considered reserved in Java).

Reply:
According to the [oracle docs](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html) '$' isn't reserved, but they do strongly recommend not using it in user names. $ is not permitted by the C standard, but GCC and Clang support it as an extension (apparently because some older compilers did), again they recommend against using it.